### PR TITLE
Fix electromagnetic field unit conversions 

### PR DESCRIPTION
**Pull request by jzuhone**
## PR Summary

Electromagnetic unit conversions in yt are complex. The reasons for this are:

* `unyt` uses MKSA as the base set of units, but `yt` assumes Gaussian CGS units by default
* MKSA has a fundamental unit for current, Ampere, which becomes a component unit of most electromagnetic units (e.g. Coulomb, Tesla, etc.)
* Gaussian CGS units do not have a fundamental unit of current, defining all electromagnetic quantities in terms of mass, length, and time units

This is described in more detail [here](https://ytep.readthedocs.io/en/master/YTEPs/YTEP-0028.html#special-handling-for-magnetic-fields). 

Some frontends assume Gaussian CGS EM units as the code units, and a few assume MKSA EM units for the code units. However, regardless of what one assumes, it ought to be straightforward and transparent to do something like this:

```python
ds = yt.load("some_dataset", unit_system="code")
```

and expect magnetic fields to have SI dimensions instead of Gaussian CGS dimensions if the underlying code units do. It turns out that this doesn't always work correctly, as seen in Issue #3471, which was originally posted about another problem but in fact led to the discovery of this issue. 

The problem is that we need something that comes even before unit detection from the dataset, to let yt know whether or not it should expect something with dimensions of SI current in it when creating the code unit system.  There is a chicken-and-egg problem where we have to set up a code unit system before we can detect units, but only when we detect units can we know whether we need an SI current unit, which is info you need to create the code unit system in the first place. It's quite awful.

This PR is an attempt to fix the logic that handles this situation, as well as the case where we have an SI code system but request a CGS system (and vice-versa), which is currently very brittle. The easiest way to do this (though clearly not the most elegant) is for the `Dataset` object to have a new attribute (called `_use_mks_em_units` in this case) which tells it to expect that the code units will be using MKSA for EM units.

I've updated the documentation and the skeleton frontend to reflect how to ensure that SI/MKSA units are used for code units, and I've also updated the stream dataset documentation to show how to make sure this happens for in-memory datasets.

I also added some new tests and fixed some other ones. Finally, this does not fix Issue #3471--that will have to be addressed in a separate PR.

## PR Checklist

- [x] New features are documented, with docstrings and narrative docs
- [x] Adds a test for any bugs fixed. Adds tests for new features.

### CODE REVIEW

File: yt/loaders.py [hunk 1]
```diff
@@ -735,6 +740,7 @@ def parse_unit(unit, dimension):
         geometry=geometry,
         unit_system=unit_system,
         default_species_fields=default_species_fields,
+        use_mks_em_units=use_mks_em_units,
```
[the remark this answers is not in the document]
Comment (jzuhone) [reply yt/loaders.py:743]: Sorry--confused. It isn't keyword-only?

[the remark this answers is not in the document]
Comment (neutrinoceros) [reply yt/loaders.py:743]: Ah, this comment was meant for the definition of `load_hexahedral_mesh`

File: yt/loaders.py [hunk 2]
```diff
@@ -754,6 +760,7 @@ def load_hexahedral_mesh(
     periodicity=(True, True, True),
     geometry="cartesian",
     unit_system="cgs",
+    use_mks_em_units=False,
```
Comment (neutrinoceros) [inline yt/loaders.py:763]: So this is what I meant to suggest early (re: keyword-only arguments)
```suggestion
    *,
    use_mks_em_units=False,
```

File: yt/loaders.py [hunk 3]
```diff
@@ -579,6 +583,7 @@ def load_particles(
     unit_system="cgs",
     data_source=None,
     default_species_fields=None,
+    use_mks_em_units=False,
```
Comment (neutrinoceros) [inline yt/loaders.py:586]: While I'm at it
```suggestion
    *,
    use_mks_em_units=False,
```

File: yt/loaders.py [hunk 4]
```diff
@@ -361,6 +363,7 @@ def load_amr_grids(
     refine_by=2,
     unit_system="cgs",
     default_species_fields=None,
+    use_mks_em_units=False,
```
Comment (neutrinoceros) [inline yt/loaders.py:366]:
```suggestion
    *,
    use_mks_em_units=False,
```

File: yt/loaders.py [hunk 5]
```diff
@@ -162,6 +162,7 @@ def load_uniform_grid(
     geometry="cartesian",
     unit_system="cgs",
     default_species_fields=None,
+    use_mks_em_units=False,
```
Comment (neutrinoceros) [inline yt/loaders.py:165]:
```suggestion
    *,
    use_mks_em_units=False,
```

File: yt/frontends/stream/data_structures.py [hunk 6]
```diff
@@ -450,13 +454,15 @@ def __init__(
         geometry="cartesian",
         unit_system="cgs",
         default_species_fields=None,
```
Comment (neutrinoceros) [inline yt/frontends/stream/data_structures.py:456]:
```suggestion
        *,
        default_species_fields=None,
```

File: yt/frontends/stream/data_structures.py [hunk 7]
```diff
@@ -274,14 +274,18 @@ def __init__(
...
+        use_mks_em_units=False,
```
Comment (neutrinoceros) [inline yt/frontends/stream/data_structures.py:277]:
```suggestion
        *,
        use_mks_em_units=False,
```

**Comment by neutrinoceros** [comment]
Aren't we supposed to wait a first approval to bump answer tests ?
Anyway I see that the one you bumped is failing again, but it's also the same flaky test that #3619 is about. If it was migrated to pytest I could propose a solution to mark it as flaky so pytest could run it more than once before degiding it failed but since it's still stuck on nose I'm really tempted to suggest we just deactivate it. That is, unless you or @matthewturk think there is a better way, of course :)

File: yt/frontends/flash/tests/test_outputs.py [hunk 8]
```diff
@@ -54,6 +57,23 @@ def test_units_override():
     units_override_check(sloshing)
 
 
+@disable_dataset_cache
```
Comment (neutrinoceros) [inline yt/frontends/flash/tests/test_outputs.py:60]: I didn't know about this decorator but after reading it, it seems to me that it's currently broken and that the previous state is *never* restored (see `disable_dataset_cache` therein). I've opened #3796 to fix this.

File: yt/frontends/stream/data_structures.py [hunk 9]
```diff
@@ -291,6 +294,25 @@ def __init__(
             default_species_fields=default_species_fields,
         )
 
+    def _check_for_mks_em_units(self):
```
Comment (neutrinoceros) [inline yt/frontends/stream/data_structures.py:297]: Doesn't hurt
```suggestion
    def _check_for_mks_em_units(self) -> bool:
```

File: yt/frontends/stream/data_structures.py [hunk 10]
```diff
@@ -291,6 +294,25 @@ def __init__(
...
+        from yt.units.dimensions import current_mks
+
+        # We assume CGS EM units by default, unless
+        # magnetic_unit is explicitly defined within MKSA
+        magnetic_unit = self.stream_handler.code_units[-1]
```
Comment (neutrinoceros) [inline yt/frontends/stream/data_structures.py:302]: I'm not too optimistic here but is there any way to make this less fragile ? If we ever append another unit to `self.stream_handler.code_units`, or change their order, this would break, possibly silently.

File: yt/frontends/_skeleton/data_structures.py [hunk 11]
```diff
@@ -114,6 +115,10 @@ def _set_code_unit_attributes(self):
         # These can also be set:
         # self.velocity_unit = self.quan(1.0, "cm/s")
         # self.magnetic_unit = self.quan(1.0, "gauss")
+        #
+        # If your frontend uses SI EM units, set magnetic units like this
+        # instead:
+        # self.magnetic_unit = self.quan(1.0, "T")
```
Comment (neutrinoceros) [inline yt/frontends/_skeleton/data_structures.py:121]: This sounds very much like a job for the `default_units` class attribute ! :)

File: yt/data_objects/static_output.py [hunk 12]
```diff
@@ -1165,25 +1185,43 @@ def _create_unit_registry(self, unit_system):
         # yt assumes a CGS unit system by default (for back compat reasons).
         # Since unyt is MKS by default we specify the MKS values of the base
         # units in the CGS system. So, for length, 1 cm = .01 m. And so on.
+        # Note that the values associated with the code units here will be
+        # modified once we actually determine what the code units are from
+        # the dataset
         self.unit_registry = UnitRegistry(unit_system=unit_system)
+        # 1 cm = 0.01 m
         self.unit_registry.add("code_length", 0.01, dimensions.length)
+        # 1 g = 0.001 kg
```
Comment (neutrinoceros) [inline yt/data_objects/static_output.py:1194]: This is actually helpful info, thank you for adding this series of comments

File: .gitignore [hunk 13]
```diff
@@ -7,6 +7,7 @@ png.cfg
 rockstar.cfg
 yt_updater.log
 yt/frontends/artio/_artio_caller.c
+yt/frontends/gamer/cfields.c
```
Comment (neutrinoceros) [inline .gitignore:10]: good catch !

**Review by neutrinoceros — COMMENTED** [review]
Okay here's a first batch of comments and questions. I have intentionally not reviewed the logic in `Dataset._assign_unit_system` yet, but I will later !

So I've checked-out the tests without the patches and I saw that only the ones from `yt/fields/tests/test_magnetic_fields.py` fail on main. Just wanted to make sure this is intentional and that the other tests your adding are intended as regression safe-guards (as opposed to being broken, as in failing to check what you wanted to check).

Also while reading this I was reminded of the existence of the (relatively new and rarely used) `Dataset.default_units` class attribute, which is intended to be subclassed. I presume the reason you didn't use it here is because you didn't know about it (can't blame you, I think only one or two frontends currently subclass it) ?

File: yt/frontends/stream/data_structures.py (hunk 10 above)
Comment (jzuhone) [reply yt/frontends/stream/data_structures.py:302]: @neutrinoceros it's not a bad idea (I thought this myself), but that's a much bigger job than this PR is trying to solve, as it would touch quite a bit of code in the stream frontend.

File: yt/units/tests/test_magnetic_code_units.py [hunk 14]
```diff
@@ -49,3 +50,25 @@ def test_magnetic_code_units():
     mucu = ds4.magnetic_unit.to("code_magnetic")
     assert_allclose(mucu.value, 1.0)
     assert str(mucu.units) == "code_magnetic"
+
+    ds5 = load_uniform_grid(
```
[the remark this answers is not in the document]
Comment (jzuhone) [reply yt/units/tests/test_magnetic_code_units.py:54]: What this test is designed to do is to cover a whole bunch of corner cases of the same phenomenon: making sure that you can load magnetic fields with any units and then convert to any other kind of magnetic field units. If one of these sub-tests is broke it's likely that something is wrong that may touch the whole machinery which they are all related to, so I would advise keeping them together.

File: yt/frontends/stream/data_structures.py (hunk 10 above)
Comment (neutrinoceros) [reply yt/frontends/stream/data_structures.py:302]: okay so I've looked into this some more and found that the code breaks (badly) in every scenario I could think of
- appending a new unit
- changing their order
- dropping the magnetic unit

So it's still not ideal, but at least I'm now confident that this will not break silently if we ever change something around it.

Comment (jzuhone) [reply yt/frontends/stream/data_structures.py:302]: We should definitely change it--I think it deserves a new PR on its own.

**Comment by jzuhone** [comment]
@neutrinoceros:

> Aren't we supposed to wait a first approval to bump answer tests ?

I wasn't aware of this, but it's been a while since I've done it.

> Anyway I see that the one you bumped is failing again, but it's also the same flaky test that https://github.com/yt-project/yt/issues/3619 is about. If it was migrated to pytest I could propose a solution to mark it as flaky so pytest could run it more than once before deciding it failed but since it's still stuck on nose I'm really tempted to suggest we just deactivate it. That is, unless you or @matthewturk think there is a better way, of course :)

This dataset and the test are related to a very botched attempt at trying to handle logarithmic grids with a variable `dds` using the semi-structured mesh machinery. It works for this dataset (but not always apparently) but often did not work for others. I'd be in favor of deactivating this test because frankly we should support datasets like this using the variable `dds` work that seems to be moving along.

File: yt/units/tests/test_magnetic_code_units.py [hunk 15]
```diff
@@ -21,16 +21,17 @@ def test_magnetic_code_units():
     assert_allclose(mucu.value, 1.0)
     assert str(mucu.units) == "code_magnetic"
 
-    ds2 = load_uniform_grid(data, ddims, magnetic_unit=(1.0, "T"), unit_system="cgs")
-
+    ds2 = load_uniform_grid(data, ddims, magnetic_unit=(1.0, "T"),
+                            unit_system="cgs")
```
[the remark this answers is not in the document]
Comment (jzuhone) [reply yt/units/tests/test_magnetic_code_units.py:25]: I edited this file in a later commit, and it's proving quite messy to back this particular change out, so I'll just fix the lines

[the remark this answers is not in the document]
Comment (neutrinoceros) [reply yt/units/tests/test_magnetic_code_units.py:25]: I was suggesting you could just squash the whole history of this branch into a single commit and leave no trace of the back-and-forth edits :)

[the remark this answers is not in the document]
Comment (jzuhone) [reply yt/units/tests/test_magnetic_code_units.py:25]: @neutrinoceros I admit I am not a big fan of that approach--I squashed a few commits but I would prefer to leave things as-is.

[the remark this answers is not in the document]
Comment (neutrinoceros) [reply yt/units/tests/test_magnetic_code_units.py:25]: that's fine, it's not a big deal here

File: yt/data_objects/static_output.py [hunk 16]
```diff
@@ -242,10 +243,10 @@ def __init__(
         self.no_cgs_equiv_length = False
 
         if unit_system == "code":
-            # create a fake MKS unit system which we will override later to
+            # create a fake unit system which we will override later to
             # avoid chicken/egg issue of the unit registry needing a unit system
             # but code units need a unit registry to define the code units on
-            used_unit_system = "mks"
+            used_unit_system = "mks" if self._use_mks_em_units else "cgs"
```
[the remark this answers is not in the document]
Comment (jzuhone) [reply yt/data_objects/static_output.py:249]: So in theory it would be nice to use `default_units` for this instead, but there is the issue that the stream frontend would still have to change `default_units` if we used it to check for MKS EM units. I'm supposing that wasn't the intent--that it should not be edited on a per-object basis?

File: yt/frontends/athena_pp/tests/test_outputs.py [hunk 17]
```diff
@@ -1,20 +1,13 @@
...
-)
+from yt.testing import assert_equal, requires_file, units_override_check
 from yt.utilities.answer_testing.framework import (
-    GenericArrayTest,
     data_dir_load,
     requires_ds,
     small_patch_amr,
 )
 
+"""
```
Comment (neutrinoceros) [inline yt/frontends/athena_pp/tests/test_outputs.py:10]: please add a comment with the issue's url so the reason why it's commented is apparent

File: tests/tests.yaml [hunk 18]
```diff
@@ -28,7 +28,7 @@ answer_tests:
     - yt/frontends/athena/tests/test_outputs.py:test_blast
     - yt/frontends/athena/tests/test_outputs.py:test_stripping
 
-  local_athena_pp_004:
+  local_athena_pp_006:
     - yt/frontends/athena_pp/tests/test_outputs.py:test_disk
```
Comment (neutrinoceros) [inline tests/tests.yaml:32]: I think we actually need to comment this line out, see
https://tests.yt-project.org/job/yt_py38_git/4630/testReport/junit/nose.failure/Failure/runTest/

File: yt/data_objects/static_output.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (jzuhone) [reply yt/data_objects/static_output.py:249]: @neutrinoceros any thoughts here?

[the remark this answers is not in the document]
Comment (neutrinoceros) [reply yt/data_objects/static_output.py:249]: Sorry I missed that.

`default_units` is indeed intended as a constant class attribute, but its job is just to provide sane defaults, instances are free to use `units_override` to replace them partially or totally.
Couldn't we migrate this specialised behaviour in the stream frontend ?

[the remark this answers is not in the document]
Comment (jzuhone) [reply yt/data_objects/static_output.py:249]: @neutrinoceros,

So `default_units` is for `units_override` only, correct? `units_override` was originally designed to provide units for datasets that weren't self-describing and didn't have any unit information in them--Athena datasets are notorious for this, for example. Because yt doesn't really work without assuming _some_ units, `units_override` provided sane defaults, so that you could at least do something with the dataset. We enabled `units_override` for all frontends, but it was always considered a solution of last resort, and if your dataset was self-describing you shouldn't even need it unless you were doing something weird.

In this use case for this PR, we need something that comes even before unit detection from the dataset, to let yt know it should expect something with dimensions of SI current in it when creating the code unit system. This has been a very tricky thing to manage  in yt while still allowing the end-user to switch to a different unit system, because there is a chicken-and-egg problem where we have to set up a code unit system before we can detect units, but only when we detect units can we know whether we need an SI current unit, which is info you need to create the code unit system in the first place. It's quite awful.

`default_units` seems like something quite different.

[the remark this answers is not in the document]
Comment (neutrinoceros) [reply yt/data_objects/static_output.py:249]: > So default_units is for units_override only, correct?

yes that was the original intention.

Ok I think I understand this problem better now. It's unfortunate that complexity is going up but I reckon it may be the only way right now. Thanks a lot for going into the details, I appreciate it !

**Comment by neutrinoceros** [comment]
I will finish my review tomorrow. I anticipate that I'll probably just approve with no further comments, but given the size of this PR, I would like a second review even if it's a bugfix, so I won't merge until then.

File: yt/data_objects/static_output.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (jzuhone) [reply yt/data_objects/static_output.py:249]: NP--I'm going to copy the explanation I gave into the main text so the context is there.

File: yt/data_objects/static_output.py [hunk 19]
```diff
@@ -1119,27 +1120,46 @@ def relative_refinement(self, l0, l1):
...
-                self.unit_registry.modify("code_magnetic", self.magnetic_unit.value)
+        # we need to determine if the requested unit system
+        # is mks-like: i.e., it has a current with the same
+        # dimensions as amperes.
+        mks_system = False
+        if unit_system != "code":
+            # if the unit system is known, we can check if it
+            # has a "current_mks" unit
+            us = unit_system_registry[str(unit_system).lower()]
+            mks_system = us.base_units[current_mks] is not None
```
Comment (neutrinoceros) [inline yt/data_objects/static_output.py:1131]: isn't `current_mks` undefined at this point ?

File: yt/data_objects/static_output.py [hunk 20]
```diff
@@ -1119,27 +1120,46 @@ def relative_refinement(self, l0, l1):
...
-                # system as an MKS-like system
-                if current_mks in self.magnetic_unit.units.dimensions.free_symbols:
-                    self.magnetic_unit = self.magnetic_unit.to("T").to("gauss")
-                # The following modification ensures that we get the conversion to
-                # cgs correct
-                self.unit_registry.modify(
-                    "code_magnetic", self.magnetic_unit.value * 0.1**0.5
-                )
-
+                if current_mks in mag_dims:
```
Comment (neutrinoceros) [inline yt/data_objects/static_output.py:1152]:
```python
else:
    if ...:
        ...
```
Looks like this could be simplified into
```python
elif:
    ...
```

**Review by neutrinoceros — COMMENTED** [review]
two more comments.

File: yt/data_objects/static_output.py (hunk 20 above)
Comment (jzuhone) [reply yt/data_objects/static_output.py]: Agreed.

File: yt/data_objects/static_output.py (hunk 19 above)
Comment (neutrinoceros) [reply yt/data_objects/static_output.py:1131]: Oh right, I think I was confusing it with `current_mks_units` while reading the diff.
I was also surprised that flake8 wouldn't catch an undef var.

**Comment by stonnes** [comment]
I took a look, and should not be an official reviewer, but I have a couple questions:
Does every frontend that uses MKS EM units now need to have the:
_use_mks_em_units = True
line hard-coded in the initial frontend definition?  And if so, will this break all the frontends that need it or did you put it in already?  Can the user set this in the yt.load call?  Not *at all* sure that this would actually be a good idea!

**Comment by neutrinoceros** [comment]
Sounds like a good opportunity to test my own understanding: I think it's always the responsibility of frontends to define this class attribute. I assume there may be some existing frontends for which it _should_ be set to True but neither John or I are familiar with them so there would be a need for follow up fixes there ?

**Comment by jzuhone** [comment]
@stonnes @neutrinoceros I'm going to revisit this--my understanding of what the code was doing got deeper the longer I looked at this, and I might have a way to fix the chicken-and-egg problem that will make for fewer changes. I'll report back later. I'll switch this to draft for now.

**Comment by jzuhone** [comment]
Closed in favor of PR #3805.
